[#12] HpItem 구현 및 플레이어 Hp UI 구현

Leekb0804 opened the pull request:
close #12 

**ItemTestScene에서 작업함**

Hp Item 구현

- HpItemEffect.cs 파일에서 회복량(_hpAmount 변수),  회복속도(_recoverySpeed 변수) 선언 및 아이템 효과 로직 구현

![pt1](https://github.com/Famec7/OpenSourceProject/assets/133445551/c9f02080-c9c1-468b-b482-4f3476457435)

`
using System.Collections;
using UnityEngine;

public class HpItemEffect : ItemEffectBase
{
    private float _hpAmount;
    private float _recoverySpeed;
    public override void Activate(PlayerData playerData)
    {
        Debug.Log("TestItemEffect activated!");     //debug

        _hpAmount = ((HpItemData)ItemData).hpAmount;
        _recoverySpeed = ((HpItemData)ItemData).recoverySpeed;
        playerData.hp += _hpAmount;
        StartCoroutine(IE_RecoveryHp(playerData));
    }

    public override void Deactivate()
    {
        Debug.Log("TestItemEffect deactivated!");   //debug
    }

    private IEnumerator IE_RecoveryHp(PlayerData playerData)
    {
        while (true)
        {
            playerData.hp += _hpAmount * Time.deltaTime * _recoverySpeed;
            yield return null;
        }
    }
}
`

Hp UI 구현

![image](https://github.com/Famec7/OpenSourceProject/assets/133445551/24c9fbe0-7564-436a-8413-3ee47755ffdc)

![blood_red_bar](https://github.com/Famec7/OpenSourceProject/assets/133445551/873dbbc0-0353-47cb-8e9b-dfa01d2cf2fb)
![blood_red_bar_background](https://github.com/Famec7/OpenSourceProject/assets/133445551/954d6d86-7043-4f76-a1c0-f755e1d97aba)

위 2개의 이미지 사용

아래는 PlayerData.hp 와 Hp UI 동기화 코드

`
using UnityEngine;
using UnityEngine.UI;

public class HpBar : MonoBehaviour
{
    public PlayerData playerData; // 플레이어 데이터를 참조
    public Image hpFillImage; // 체력을 표시할 이미지

    private float maxHp; // 최대 체력

    void Start()
    {
        if (playerData != null)
        {
            maxHp = playerData.hp; // 게임 시작 시 최대 체력을 설정
        }
    }

    void Update()
    {
        if (playerData != null && hpFillImage != null)
        {
            // 체력 비율에 따라 이미지의 fillAmount를 조정
            hpFillImage.fillAmount = playerData.hp / maxHp;
        }
    }
}

`

Leekb0804 committed:
feat: Hp Item 구현 완료

1. HpItemData.cs 작성
2. HpItemEffect.cs 작성
3. HpItem을 Prefab으로 생성
4. UI의 canvas에 HpBarFill, HpBarBackground을 Prefab으로 생성

Leekb0804 committed:
docs: .gitignore 파일 수정

+ .vscode 디렉토리 삭제

Famec7 commented:
```
while (true)
{
        playerData.hp += _hpAmount * Time.deltaTime * _recoverySpeed;
        yield return null;
}
```
위 코드에서 break 조건이 없어서 무한 루프 상태가 됩니다 playerData.hp가 어느정도 이상 회복이 되면 break되도록 설정해주세요

- #5 코드에서 duration이 0이면 아이템이 바로 deactive되는 버그가 있어 새 issue로 올려서 고치도록 하겠습니다

Leekb0804 commented:
확인했습니다.
반영하겠습니다.

Famec7 commented:
- [x]  #4 에서 duration이 0일 때 케이스를 나누는 것보다 회복 시간을 n초에 걸쳐서 할 수 있도록 시스템을 바꿔보는 것도 좋을 것 같습니다. duration은 hpAmount / recoverySpeed로 조정하면 좋을 것 같습니다.

- [x] activate함수에 playerData.hp+=hpAmount는 지워야 합니다. Coroutine에서 이미 hp를 회복하고 있습니다.
- [x] PR시 자동으로 build가 되어서 Debug.Log함수는 지우거나 전처리기를 사용하면 좋을 것 같습니다. 전처리기 사용법은 #4 를 확인해보면 좋습니다.
- [x] 자동으로 build가 되기때문에 buildSetting을 테스트할 Scene으로 바꾸면 좋을 것 같습니다.

Leekb0804 committed:
refactor: HpItemEffect 수정

1. Activate 에 중복되는 playerData.hp += _hpAmount; 삭제
2. Activate 에 duration 설정
3. Deactivate 에 Debug.log 삭제

Leekb0804 committed:
fix: 플레이어 회복 로직 수정

1. finalHp = _hpAmount + playerData.hp;
    while(playerData.hp < finalHp)
     playerData.hp += _recoverySpeed * Time.deltaTime;

Leekb0804 committed:
refactor: 주석 추가 및 debug.log 삭제

1. Activate 의 debug.log 삭제
2. HpItemEffect.cs와 HpItemData.cs 에 주석 추가

Famec7 committed:
Merge branch 'main' into feature/hp-potion-item

Famec7 reviewed (commented):
확인했습니다 MaxHp와 Item 스폰 기능 부분은 #23 이 해결되어야 가능할 거 같아서 이후 Issue로 추가하겠습니다

Famec7 reviewed (approved):
확인했습니다 MaxHp와 Item 스폰 기능 부분은 #23 이 해결되어야 가능할 거 같아서 이후 Issue로 추가하겠습니다
